pysys/process/user.py

```diff
@@ -189,8 +198,175 @@ def startProcess(self, command, arguments, environs=None, workingDir=None, state
+
+		If needed this method can be overridden in subclasses to add common 
+		environment variables for every process invoked by startProcess, for 
+		example to enable options such as code coverage for Java/Python/etc. 
+		This is also a good place to customize behaviour for different 
+		operating systems. 
+
+		Some features of this method can be configured by setting project 
+		properties:
+		
+		  - `defaultEnvironsDefaultLang`: if set to a value such as `en_US.UTF-8` 
+		    the specified value is set for the LANG= variable on Unix; otherwise, 
+		    the LANG variable is not set (which might result in use of the 
+		    legacy POSIX/C encoding).
+		  
+		  - `defaultEnvironsTempDir`: if set the expression will be passed to 
+		    Python `eval()` and used to set the OS-specific temp directory 
+		    environment variables. A typical value is `self.output`.
+		
+		  - `defaultEnvironsLegacyMode`: set to true to enable compatibility 
+		    mode which keeps the behaviour the same as PySys v1.1, 1.2 and 1.3, 
+		    namely using a completely empty default environment on Unix, and 
+		    a copy of the entire parent environment on Windows. This is not 
+		    recommended unless you have a lot of legacy tests that cannot 
+		    easily be changed to only set minimal required environment 
+		    variables using `createEnvirons()`. 
+
+		@param command: If known, the full path of the executable for which 
+		a default environment is being created (when called from `startProcess` 
+		this is always set). This allows default environment variables to be 
+		customized for different process types e.g. Java, Python, etc. 
+		
+		@param kwargs: Overrides of this method should pass any additional 
+		kwargs down to the super implementation, to allow for future extensions. 
+		
+		@return: A new dictionary containing the environment variables. 
+		"""
+		
+		assert not kwargs, 'Unknown keyword arguments: %s'%kwargs.keys()
+
+		# this feature is a workaround to maintain compatibility for a bug in PySys v1.1-1.3
+		# (see https://github.com/pysys-test/pysys-test/issues/9 for details)
+		if getattr(PROJECT, 'defaultEnvironsLegacyMode','').lower()=='true':
+			if IS_WINDOWS: 
+				return dict(os.environ)
+			else:
+				return {}
+
+		e = {}
+
+		# allows setting TEMP to output dir to avoid contamination/filling up of system location
+		if getattr(PROJECT, 'defaultEnvironsTempDir',None)!=None:
+			tempDir = eval(PROJECT.defaultEnvironsTempDir)
+			self.mkdir(tempDir)
+			if IS_WINDOWS: # pragma: no cover
+				e['TEMP'] = e['TMP'] = os.path.normpath(tempDir)
+			else:
+				e['TMPDIR'] = os.path.normpath(tempDir)
+	
+		inherited = [] # env vars where it is safe and useful to inherit parent values
+		if IS_WINDOWS: 
+			# for windows there are lots; common programs such as Python fail if these are missing
+			inherited.extend(['ComSpec', 'OS', 'PATHEXT', 'SystemRoot', 'SystemDrive', 'windir', 'NUMBER_OF_PROCESSORS'])
```

Comment:
I know you're trying to keep the default environment minimal but I also wouldn't want to get into a situation where this list has to be repeatedly tweaked because important things turn out to be missing.  So I'm wondering if we need to be a bit more rigorous than just including the envars that allowed Python to start up?  Need to think more about how we would actually do that but a couple of initial thoughts that occur to me are:

- Should any of the `Program*` and `Common*` variables be included?  I can imagine programs using these to find install/data directories and failing to work without them.
- Are there any differences between domain and non-domain machines that should be accounted for?
- If `NUMBER_OF_PROCESSORS` is in why not `PROCESSOR_*`?

Reply:
Yep it's a good question. 

Since there's no standard list or definitive reference source unfortunately, the current list is the product of me working through the Windows env vars and adding the ones that seemed to have a good chance of being needed, partly informed by experience with python. I'm very open to suggestions of extra things that you think should be added, if they're safe (i.e. don't have a likely effect of making tests depend on or write to areas outside of pysys's control such as user's home area) - specific answers to your specific suggestsion are at the bottom of this comment. 

Microsoft doesn't seem to have a standard list of env var and of course the set will vary by OS as well. The closest I've found is https://docs.microsoft.com/en-us/windows/deployment/usmt/usmt-recognized-environment-variables but it doesn't give detail om which OS versions use each one. And of course it doesn't give a lot of clues on how much each one is used by programs in the real world. And if we expanded that short list to be closer to the 30-40 or so on a default windows machine it'd be unwieldy to maintain and introduce an expectation among PySys users that "almsot everything they mgiht possibly need" from the environment will be there. I'd rather set an expectation that almost nothing is there unless you explicitly add it, whcih is easier for us to live up to and avoids disappointing people. 

Users who are gung-ho about impact of current user's environment on their tests or jsut want to "get it working quickly" already have the very simple option of just passing environs=os.environ (i.e. the entire parent environment) so I feel that base is already covered pretty well - whereas these helper methods are there to help more conservative users who are leaning the other way. 

So policy-wise I'd rather err on the side of keeping the default list too small rather than too large, and then let individual tests and/or custom basetest helper methods selectively add extra env vars as needed for a given process than make it too large and risk introducing unwanted test contamination. 

One of the most tricky bits is per-user config. I suspect a good chunk of programs will rely on user-specific env vars like USERNAME, APPDATA etc which I don't want to pass in automatically since it increases the chance of unexpected test contamination/interaction with areas outside the test output dir. 

> Should any of the Program* and Common* variables be included? I can imagine programs using these to find install/data directories and failing to work without them.

Yeah you've persuaded me on that - I'll add them. 

> Are there any differences between domain and non-domain machines that should be accounted for?

This is sort of minefield we'd get into if we aimed to provide a near-complete set of env vars. I think keeping it minimal frees us from the need to navigate that minefield (and revisit it for every old and new windows version, which we don't have the QA bandwidth for). 

> If NUMBER_OF_PROCESSORS is in why not PROCESSOR_*?
 
OK I'll add PROCESSOR_ARCHITECTURE. However I really don't beleive many apps are liekly to rely on "PROCESSOR_IDENTIFIER=Intel64 Family 6 Model 158 Stepping 10, GenuineIntel" so it think it fails the usefulness test. As mentioned, if someone does need it they can us os.environ or add it.

Reply:
Yep, that seems reasonable enough.